lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -92,9 +92,36 @@ static bool isArrayLiteralIntrinsic(ApplyInst *ai) {
 }
 
 static ApplyInst *getAllocateUninitializedArrayIntrinsic(SILValue v) {
-  if (auto *applyInst = dyn_cast<ApplyInst>(v))
-    if (isArrayLiteralIntrinsic(applyInst))
-      return applyInst;
+  if (auto *ai = dyn_cast<ApplyInst>(v))
+    if (isArrayLiteralIntrinsic(ai))
+      return ai;
+  return nullptr;
+}
+
+/// Given an element address from an `array.uninitialized_intrinsic` `apply`
```

Comment:
Do literals with tuples get handled properly? For example, `[(x, y)]` gets initialized by extracting the addresses of the tuple elements and then storing `x` and `y` to them.

To handle that, this might need to understand tuple_element_addr instructions.

I think a similar thing might also happen for arrays of structs, though I couldn't trigger it with simple examples, so maybe it doesn't. Seems like it might be safer to handle all address extraction instructions anyways.

Reply:
Great observation.  I don't think this will occur in reality though, because tuples cannot conform to protocols. Even if you make a custom type be `ExpressibleByArrayLiteral` where `ArrayLiteralElement == (Float, Float)`, there still doesn't seem to be any code path that would trigger it, because `init(arrayLiteralElements:)` cannot be marked `@differentiable`.

Reply:
Oh, seems like the tuple case is safeish! Maybe there is some struct case that will break it? I tried a few simple examples, and I couldn't make an array literal initialization that uses struct_element_addr, but I feel like it might happen sometimes.